pkg/controller/version/docker.go

```diff
@@ -24,62 +27,87 @@ func NewDockerVersionChecker(currentImage, currentImageId string, dockerConfig *
 }
 
 func (dockerVersionChecker *DockerVersionChecker) IsLatest() (bool, error) {
-	regex := regexp.MustCompile("(^docker-pullable:\\/\\/|\\:.*$|\\@sha256.*$)")
-	latestImageName := regex.ReplaceAllString(dockerVersionChecker.currentImage, "") + ":latest"
+	//regex := regexp.MustCompile("(^docker-pullable:\\/\\/|\\:.*$|\\@sha256.*$)")
+	//latestImageName := regex.ReplaceAllString(dockerVersionChecker.currentImage, "") + ":latest"
 
-	//Using ImageID instead of Image because ImageID contains digest of image that is used while Image only contains tag
-	reference, err := name.ParseReference(strings.TrimPrefix(dockerVersionChecker.currentImageId, "docker-pullable://"))
+	transportImageName := fmt.Sprintf("%s%s",
+		"docker://",
+		strings.TrimPrefix(
+			dockerVersionChecker.currentImageId,
+			"docker-pullable://"))
+
+	latestReference, err := alltransports.ParseImageName("docker://" + dockerVersionChecker.currentImage)
 	if err != nil {
 		return false, err
 	}
 
-	latestReference, err := name.ParseReference(latestImageName)
+	latestDigest, err := dockerVersionChecker.getDigest(latestReference)
 	if err != nil {
 		return false, err
 	}
 
-	registryURL := "https://" + reference.Context().RegistryStr()
-	authOption := getAuthOption(dockerVersionChecker.dockerConfig, registryURL)
-
-	latestDigest, err := getDigest(latestReference, authOption)
+	//Using ImageID instead of Image because ImageID contains digest of image that is used while Image only contains tag
+	currentReference, err := alltransports.ParseImageName(transportImageName)
+	//reference, err := name.ParseReference(strings.TrimPrefix(dockerVersionChecker.currentImageId, "docker-pullable://"))
 	if err != nil {
 		return false, err
 	}
 
-	currentDigest, err := getDigest(reference, authOption)
+	currentDigest, err := dockerVersionChecker.getDigest(currentReference)
 	if err != nil {
 		return false, err
 	}
 
-	return currentDigest == latestDigest, nil
+	return currentDigest.Algorithm().String() == latestDigest.Algorithm().String() &&
+		currentDigest.Hex() == latestDigest.Hex(), nil
 }
 
-func getDigest(reference name.Reference, authOption remote.Option) (string, error) {
-	img, err := remote.Image(reference, authOption)
+func (dockerVersionChecker *DockerVersionChecker) getDigest(ref types.ImageReference) (*digest.Digest, error) {
+	systemContext := dockerVersionChecker.makeSystemContext(ref.DockerReference())
+	imageSource, err := ref.NewImageSource(context.TODO(), systemContext)
+	defer closeImageSource(imageSource)
```

Comment:
Can `imageSource` be set when there is an error? Otherwise makes sense to defer the `imageSource.Close()` directly after the `if`.